### PR TITLE
Fixes #3169 issue with selectOnClose errors

### LINKED ISSUE #3169: selectOnClose stack overflow
When I have SelectOnClose = true; I am recieving this error: I am using the most recent version of 4rc2.

```
Uncaught SyntaxError: Invalid regular expression: /(^|\.)select2\.(?:.*\.|)undefined(\.|$)/: Stack overflow		jQuery 
b.event.remove		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.off		jSelect2.
S2.define.AttachBody._detachPositioningHandler		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
(anonymous function)		jQuery 
b.event.dispatch		jQuery 
b.event.add.v.handle		jQuery 
b.event.trigger		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.trigger		jSelect2.
S2.define.SelectOnClose._handleSelectOnClose		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
(anonymous function)		jQuery 
b.event.dispatch		jQuery 
b.event.add.v.handle		jQuery 
b.event.trigger		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.trigger		jSelect2.
S2.define.SelectOnClose._handleSelectOnClose		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
(anonymous function)		jQuery 
b.event.dispatch		jQuery 
b.event.add.v.handle		jQuery 
b.event.trigger		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.trigger		jSelect2.
S2.define.SelectOnClose._handleSelectOnClose		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
(anonymous function)		jQuery 
b.event.dispatch		jQuery 
b.event.add.v.handle		jQuery 
b.event.trigger		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.trigger		jSelect2.
S2.define.SelectOnClose._handleSelectOnClose		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
(anonymous function)		jQuery 
b.event.dispatch		jQuery 
b.event.add.v.handle		jQuery 
b.event.trigger		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.trigger		jSelect2.
S2.define.SelectOnClose._handleSelectOnClose		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
(anonymous function)		jQuery 
b.event.dispatch		jQuery 
b.event.add.v.handle		jQuery 
b.event.trigger		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.trigger		jSelect2.
S2.define.SelectOnClose._handleSelectOnClose		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
(anonymous function)		jQuery 
b.event.dispatch		jQuery 
b.event.add.v.handle		jQuery 
b.event.trigger		jQuery 
(anonymous function)		jQuery 
b.extend.each		jQuery 
b.fn.b.each		jQuery 
b.fn.extend.trigger		jSelect2.
S2.define.SelectOnClose._handleSelectOnClose		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.Select2.trigger		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke		jSelect2.
S2.define.Observable.trigger		jSelect2.
S2.define.CloseOnSelect._selectTriggered		jSelect2.
S2.define.Utils.Decorate.calledMethod		jSelect2.
(anonymous function)		jSelect2.
S2.define.Observable.invoke
```

### PR DESCRIPTION
**Pull request by Brankub**
Fix related issue: https://github.com/select2/select2/issues/3169

### CODE REVIEW

File: src/js/select2/dropdown/selectOnClose.js [hunk 1]
```diff
@@ -8,19 +8,24 @@ define([
...
+
     var $highlightedResults = this.getHighlightedResults();
 
     if ($highlightedResults.length < 1) {
       return;
     }
 
-    $highlightedResults.trigger('mouseup');
+    self.trigger('select', {
+        originalEvent: evt,
```
Comment (kevin-brown) [inline src/js/select2/dropdown/selectOnClose.js]: We don't need to pass `originalEvent` along here, as `evt` isn't a DOM event.

Comment (Brankub) [inline src/js/select2/dropdown/selectOnClose.js]: Thank you for your comment. I removed it in next commit.

File: src/js/select2/dropdown/selectOnClose.js [hunk 2]
```diff
@@ -14,13 +14,17 @@ define([
...
+    var self = this;
+
     var $highlightedResults = this.getHighlightedResults();
 
     if ($highlightedResults.length < 1) {
       return;
     }
 
-    $highlightedResults.trigger('mouseup');
+    self.trigger('select', {
```
Comment (kevin-brown) [inline src/js/select2/dropdown/selectOnClose.js]: This can also be switched back to `this`, we don't need `self` anymore.

**Comment by kevin-brown** [comment]
Thanks for the commits. I've left one last inline comment and then I think this should be all set to be merged in.

File: src/js/select2/dropdown/selectOnClose.js (hunk 2 above)
Comment (Brankub) [inline src/js/select2/dropdown/selectOnClose.js]: you are right, swtiched back

**Comment by Brankub** [comment]
Thank you for your comments, hope this fix would be helpful.

**Comment by kevin-brown** [comment]
Thanks for making the changes, this should fix the issue for the forseeable future.

**Pull request merged**